### PR TITLE
docs: add re-delegation design document

### PR DESCRIPTION
**Pull request by sveitser**
Describe contract changes (StakeTableV3), Rust event handling, CLI commands, delegation UI changes, upgrade strategy, and forward compatibility with slashing. Include test plan with requirements, edge cases, withdrawal path exclusion tests, UI tests, and invariant test specifications.

### CODE REVIEW

File: doc/stake-redelegation-feature.md [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+### Files to change
+
+**`types/src/v0/v0_3/stake_table.rs`**
+
+- Add `Redelegate` variant to `StakeTableEvent`
+
+**`types/src/v0/impls/stake_table.rs`**
+
+- Rust bindings must be regenerated with the V3 ABI (`just gen-bindings`) so that `StakeTableV2Events` (or a new
+  `StakeTableV3Events`) includes the `Redelegated` variant. Without this, `decode_raw_log` silently drops the event.
```
[the remark this answers is not in the document]
Comment (sveitser) [reply doc/stake-redelegation-feature.md]: /gemini check

File: doc/stake-redelegation-feature.md [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+
+- `TryFrom` impl (~line 129): map `Redelegated` filter log to
+  `StakeTableEvent::Redelegate { delegator, from_validator, to_validator, amount }`
+
+- `apply_event` (~line 299): handle `Redelegate` as an atomic move:
+  - Subtract `amount` from source validator's delegated stake and delegator's delegation
+  - Add `amount` to destination validator's delegated stake and delegator's delegation
+  - Pattern matches existing `Delegate` (line 418) and `Undelegate` (line 459) handlers
```
[the remark this answers is not in the document]
Comment (sveitser) [reply doc/stake-redelegation-feature.md]: /gemini check

File: doc/stake-redelegation-feature.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sveitser) [reply doc/stake-redelegation-feature.md]: /gemini check

File: doc/stake-redelegation-feature.md [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+with V2.
+
+### `initializeV3()`
+
+Minimal. No new storage to initialize.
+
+### New function
+
+```solidity
+function redelegate(
```
Comment (Ayiga) [inline doc/stake-redelegation-feature.md:44]: The name `redelegate`, to me, implies that we are adding an additional delegation to an existing `validator` to whom we are already delegated. 

It seems that the purpose of this is to effectively transfer a delegation from one validator to another.  Perhaps we could have a name that is more representative of that action?

`moveDelegation`?

File: doc/stake-redelegation-feature.md [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+
+## 5. Delegation UI Changes
+
+**Repo:** `espresso-block-explorer` (`packages/espresso-block-explorer-components`)
+
+### Contract interface
+
+Add `redelegate` to the StakeTable contract interface and ABI. Follows the existing `delegate`/`undelegate` pattern.
+
+**Files:**
```
Comment (Ayiga) [inline doc/stake-redelegation-feature.md:147]: In the project, I have separated out the `stake_table` implementations / interfaces based on the contract version.

If I were to keep this up, I would probably end up creating `contracts/stake_table_v3` which would extend `contracts/stake_table_v2`.

File: doc/stake-redelegation-feature.md [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+- `contracts/stake_table_v2/stake_table_v2_interface.ts` — extend V2 writeable interface with
+  `redelegate(fromValidator, toValidator, amount)` (mirrors Solidity inheritance: `StakeTableV3 is StakeTableV2`)
+- `contracts/stake_table_v2/stake_table_v2_abi.ts` — add `redelegate` function and `Redelegated` event ABI entries
+
+### State machine
+
+Add new states to `ValidatorSelectionContext` (`validator_selection_context.tsx`):
+
+- `ValidatorConfirmedRedelegate` — user chose "Redelegate" from manage stake. Carries source `validatorAddress`. Shows
+  amount input with "Max" button (auto-fills full delegation to avoid dust).
```
Comment (Ayiga) [inline doc/stake-redelegation-feature.md:158]: "dust"?

File: doc/stake-redelegation-feature.md [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+- New `perform_redelegation_context.tsx` — async generator context following the `performDelegation` /
+  `performUndelegation` pattern.
+- Calls `stakeTableContract.redelegate(fromValidator, toValidator, amount)`.
+- On success, triggers L1 data refresh via `SetL1RefreshTimestampContext` (existing pattern).
+- On wallet rejection (user cancels signing), return to confirm screen without error banner.
+- No new data fetching needed — existing `WalletSnapshot` tracks delegations per validator.
+
+### Event display
+
+`Redelegated` events should appear in the delegation history / activity feed alongside `Delegated` and `Undelegated`
```
Comment (Ayiga) [inline doc/stake-redelegation-feature.md:197]: We don't have an activity feed currently.

File: doc/stake-redelegation-feature.md (hunk 5 above)
Comment (sveitser) [reply doc/stake-redelegation-feature.md:158]: very small amounts

File: doc/stake-redelegation-feature.md (hunk 3 above)
Comment (sveitser) [reply doc/stake-redelegation-feature.md:44]: It was the term used by product. We can discuss on slack. It might be a commonly accepted term in staking lingo.

File: doc/stake-redelegation-feature.md [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+
+1. `undelegate(fromValidator, amount)` -- begins escrow
+2. Wait for escrow period to expire
+3. `claimWithdrawal()` -- tokens returned to delegator
+4. `delegate(toValidator, amount)` -- stake with new validator
+
+This locks capital during the escrow period, penalizing delegators who want to rebalance.
+
+**Redelegation** replaces this with a single `redelegate(fromValidator, toValidator, amount)` transaction. Tokens move
+atomically between validators with no escrow, no capital lockup, and no change to `activeStake`.
```
Comment (alysiahuggins) [inline doc/stake-redelegation-feature.md:15]: since redelegation is instant and atomic, one concern i have is a coordinated redelegation of stake that causes a validator to become the majority stake holder (above 1/3) in a few mins. Perhaps there can be a per operator stake cap in the contract. However, this concern may be acceptable at the network's current maturity level

File: doc/stake-redelegation-feature.md [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+`Redelegated` events should appear in the delegation history / activity feed alongside `Delegated` and `Undelegated`
+events. Display: "Redelegated {amount} from {sourceValidator} to {destValidator}".
+
+## 6. Upgrade Strategy
+
+V3 contract deployment with operational coordination:
+
+1. Release new sequencer binary with `Redelegate` event support (dormant - no contract emits it yet)
+2. All validators upgrade to new binary
+3. Verify all (or almost all) validators are running the new binary (via metrics API endpoints or operator coordination)
```
Comment (alysiahuggins) [inline doc/stake-redelegation-feature.md:206]: if all validators don't upgrade, what happens if some drop the redelegated event and then have a diff view of the stake table?

Comment (sveitser) [reply doc/stake-redelegation-feature.md:206]: Assuming its only a small part of the stake table those validors would no longer participate in the majority consensus protocol until they upgrade. So in practice I we would just get more timeouts/missed slots until they upgrade. It's the same for protocol upgrades we require a 90% stake majority to upgrade, not 100%.

Comment (alysiahuggins) [reply doc/stake-redelegation-feature.md:206]: ah so like a soft fork, got it
